src/test/java/org/junit/platform/engine/test/event/ExecutionEventRecorder.java

```diff
@@ -0,0 +1,175 @@
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.SKIPPED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.STARTED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byPayload;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byTestDescriptor;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byType;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.EngineExecutionListener;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.TestExecutionResult.Status;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.engine.test.event.ExecutionEvent.Type;
+
+/**
+ * {@link EngineExecutionListener} that records all events and makes them available to tests.
+ *
+ * @since 1.0
+ * @see ExecutionEvent
+ */
+public class ExecutionEventRecorder implements EngineExecutionListener {
+
+	public static List<ExecutionEvent> execute(TestEngine testEngine, EngineDiscoveryRequest discoveryRequest) {
+		TestDescriptor engineTestDescriptor = testEngine.discover(discoveryRequest,
+			UniqueId.forEngine(testEngine.getId()));
+		ExecutionEventRecorder listener = new ExecutionEventRecorder();
+		testEngine.execute(
+			new ExecutionRequest(engineTestDescriptor, listener, discoveryRequest.getConfigurationParameters()));
+		return listener.getExecutionEvents();
+	}
+
+	public final List<ExecutionEvent> executionEvents = new CopyOnWriteArrayList<>();
+
+	@Override
+	public void dynamicTestRegistered(TestDescriptor testDescriptor) {
+		addEvent(ExecutionEvent.dynamicTestRegistered(testDescriptor));
+	}
+
+	@Override
+	public void executionSkipped(TestDescriptor testDescriptor, String reason) {
+		addEvent(ExecutionEvent.executionSkipped(testDescriptor, reason));
+	}
+
+	@Override
+	public void executionStarted(TestDescriptor testDescriptor) {
+		addEvent(ExecutionEvent.executionStarted(testDescriptor));
+	}
+
+	@Override
+	public void executionFinished(TestDescriptor testDescriptor, TestExecutionResult result) {
+		addEvent(ExecutionEvent.executionFinished(testDescriptor, result));
+	}
+
+	@Override
+	public void reportingEntryPublished(TestDescriptor testDescriptor, ReportEntry entry) {
+		addEvent(ExecutionEvent.reportingEntryPublished(testDescriptor, entry));
+	}
+
+	public List<ExecutionEvent> getExecutionEvents() {
+		return executionEvents;
+	}
+
+	public Stream<ExecutionEvent> eventStream() {
+		return getExecutionEvents().stream();
+	}
+
+	public long getTestSkippedCount() {
```

Comment:
The underlying stream is created from `executionEvents`, which being a `List` cannot have more than `Integer.MAX_VALUE` elements.

Have you considered casting the result of `.count()` to an `int` and making this method and all the other `long`-returning methods below return `int` instead?